py/runtime.c

```diff
@@ -1409,3 +1409,76 @@ NORETURN void mp_raise_TypeError(const char *msg) {
 NORETURN void mp_not_implemented(const char *msg) {
     mp_raise_msg(&mp_type_NotImplementedError, msg);
 }
+
+#if MICROPY_PY_SOFTIRQ
+
+#define MAX_STACK_DEPTH (5)
+
+typedef struct _stack_elem_t {
+    mp_obj_t irqfunc;
+    mp_obj_t arg1;
+    mp_obj_t arg2;
+} stack_elem_t;
+
+volatile bool softirq_enabled = false;
+volatile short sp = -1;
+stack_elem_t stack[MAX_STACK_DEPTH];
+
+void mp_enable_softirq(void) {
+    softirq_enabled = true;
+}
+
+void mp_disable_softirq(void) {
+    mp_uint_t irqstate = disable_irq();
+    softirq_enabled = false;
+    sp = -1;
+    MP_STATE_VM(mp_pending_ex_flags) &= ~PENDING_EX_SOFT_INT;
+    enable_irq(irqstate);
+}
+
+void mp_exec_softirq(void) {
+    stack_elem_t softirq;
+    mp_uint_t irqstate;
+    static bool busy = false;
+
+    if (sp >= 0 && busy == false && __get_IPSR() == 0) {
+        irqstate = disable_irq();
+        softirq = stack[sp--];
+        busy = true;
+        enable_irq(irqstate);
+        if (softirq.arg1 == MP_OBJ_NULL) {
+            mp_call_function_0(softirq.irqfunc);
+        } else if (softirq.arg2 == MP_OBJ_NULL) {
+            mp_call_function_1(softirq.irqfunc, softirq.arg1);
+        } else {
+            mp_call_function_2(softirq.irqfunc, softirq.arg1, softirq.arg2);
+        }
+        irqstate = disable_irq();
+        busy = false;
+        if (sp == -1) {
+            MP_STATE_VM(mp_pending_ex_flags) &= ~PENDING_EX_SOFT_INT;
+        }
+        enable_irq(irqstate);
+    }
+}
+
+mp_obj_t mp_add_softint(mp_obj_t function, mp_obj_t arg1, mp_obj_t arg2) {
+
+    if (sp < MAX_STACK_DEPTH - 1 && softirq_enabled) {
```

Comment:
Per object, or per underlying HwIRQ? If per object, that may require quite a bunch of space, consider for example 200 GPIOs, for each 2 args stored. Using a queue is more frugal way to deal with that.

> do they fill up the queue? 

Yes, and being dropped, just the same as real IRQs (sometimes).

Generally, this question is about edge- vs level- based interrupt handling. HwIRQ can be both, but what should be SoftIRQ? Edge-based solution is more real-world usable, if not say only real-world usable: if IRQ happened, that fact will be recorded, and thus there will be chance to serve it, even if other interrupts are serviced for a long time, or just disabled. Actually, how level interrupts are usually used is by emulation edge-based scheme, except the IRQ state is latched in external device, instead of CPU.

And what @HenrikSolver implemented is more like level-based scheme: if softirqs are "disabled", they aren't recorded at all. Except that his disable soft IRQ isn't real disable (in the sense of HardIRQ), but tailored specifically to not process SoftIRQ while in REPL (not a desired behavior). And note that on a real hardware, if interrupts disabled, level interrupt asserted, interrupts enabled and level interrupt is still asserted, it will trigger handler. But it's hard to say what "duration" pending softirqs have, so faithfully emulating level-based scheme in software is complicated, so another reason to go for (faithful) edge-based scheme.